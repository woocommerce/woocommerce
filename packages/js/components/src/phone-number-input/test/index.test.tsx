/**
 * External dependencies
 */
import { createElement } from '@wordpress/element';
import { noop } from 'lodash';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

/**
 * Internal dependencies
 */
import PhoneNumberInput from '..';

describe( 'PhoneNumberInput', () => {
	it( 'should match snapshot', () => {
		const { container } = render(
			<PhoneNumberInput value="" onChange={ noop } />
		);
		expect( container ).toMatchSnapshot();
	} );

	it( 'should match snapshot with custom renders', () => {
		const { container } = render(
			<PhoneNumberInput
				value=""
				onChange={ noop }
				selectedRender={ ( { name } ) => name }
				itemRender={ ( { code } ) => code }
				arrowRender={ () => '⬇️' }
			/>
		);
		expect( container ).toMatchSnapshot();
	} );

	it( 'should render with provided `id`', () => {
		render( <PhoneNumberInput id="test-id" value="" onChange={ noop } /> );
		expect( screen.getByRole( 'textbox' ) ).toHaveAttribute(
			'id',
			'test-id'
		);
	} );

	it( 'calls onChange callback on number input', () => {
		const onChange = jest.fn();
		render( <PhoneNumberInput value="" onChange={ onChange } /> );

		const input = screen.getByRole( 'textbox' );
		userEvent.type( input, '1' );

		expect( onChange ).toHaveBeenCalledWith( '+1 1', '+11', 'US' );
	} );

	it( 'calls onChange callback when a country is selected', () => {
		const onChange = jest.fn();
		render( <PhoneNumberInput value="0 0" onChange={ onChange } /> );

		const select = screen.getByRole( 'button' );
		userEvent.click( select );

		const option = screen.getByRole( 'option', { name: /es/i } );
		userEvent.click( option );

		expect( onChange ).toHaveBeenCalledWith( '+34 0 0', '+3400', 'ES' );
	} );

	it( 'prevents consecutive spaces and hyphens', () => {
		const onChange = jest.fn();
		render( <PhoneNumberInput value="0-" onChange={ onChange } /> );

		const input = screen.getByRole( 'textbox' );
		userEvent.type( input, '-' );

		expect( onChange ).not.toHaveBeenCalled();
	} );
} );
