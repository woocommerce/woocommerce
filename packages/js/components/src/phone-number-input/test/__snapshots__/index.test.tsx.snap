// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhoneNumberInput should match snapshot 1`] = `
<div>
  <div
    class="wcpay-component-phone-number-input"
  >
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-labelledby="downshift-0-label downshift-0-toggle-button"
      class="wcpay-component-phone-number-input__button"
      id="downshift-0-toggle-button"
      type="button"
    >
      <img
        alt="US flag"
        class="wcpay-component-phone-number-input__flag"
        src="https://s.w.org/images/core/emoji/14.0.0/72x72/1f1fa-1f1f8.png"
      />
       +1
      <span
        class="wcpay-component-phone-number-input__button-arrow"
      >
        <svg
          aria-hidden="true"
          focusable="false"
          height="18"
          viewBox="0 0 24 24"
          width="18"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M17.5 11.6L12 16l-5.5-4.4.9-1.2L12 14l4.5-3.6 1 1.2z"
          />
        </svg>
      </span>
    </button>
    <input
      class="wcpay-component-phone-number-input__input"
      style="padding-left: 0px;"
      type="text"
      value=""
    />
    <ul
      aria-hidden="true"
      aria-labelledby="downshift-0-label"
      class="wcpay-component-phone-number-input__menu"
      id="downshift-0-menu"
      role="listbox"
      tabindex="-1"
    />
  </div>
</div>
`;

exports[`PhoneNumberInput should match snapshot with custom renders 1`] = `
<div>
  <div
    class="wcpay-component-phone-number-input"
  >
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-labelledby="downshift-1-label downshift-1-toggle-button"
      class="wcpay-component-phone-number-input__button"
      id="downshift-1-toggle-button"
      type="button"
    >
      US
      <span
        class="wcpay-component-phone-number-input__button-arrow"
      >
        ⬇️
      </span>
    </button>
    <input
      class="wcpay-component-phone-number-input__input"
      style="padding-left: 0px;"
      type="text"
      value=""
    />
    <ul
      aria-hidden="true"
      aria-labelledby="downshift-1-label"
      class="wcpay-component-phone-number-input__menu"
      id="downshift-1-menu"
      role="listbox"
      tabindex="-1"
    />
  </div>
</div>
`;
